docs/proposals/ose-rebase.md

```diff
@@ -0,0 +1,105 @@
+# will drop it and tito tag needs it further down the build process in order to determine
+# the next (new) tag.
+last_tag="$( git describe --abbrev=0 --tags )"
+
+# sanity check that ensures the last three commits are in the right order.
+#
+# The last three commits are the tito tag commit, a web console bump that got created
+# during the previous rebase, and the specfile updates:
+#
+#         .-Other [CARRY] commit
+#        /  .-[CARRY] Specfile updates
+#       /  /  .-[DROP] webconsole bump
+#      /  /  /  .-Tito tag commit
+#     /  /  /  /
+# m--c--c--w--t
+#  \           \
+#   \           `-ose/master/HEAD,v3.5.0.23
+#    `-origin/master/HEAD
+#
+sanity_check
+
+# Reset the last three commits.
+git reset HEAD~3
+
+# Squash the tito commit into the specfile updates
+git add .tito origin.spec
+git commit -m "[CARRY][BUILD] Specfile updates"
+
+# Drop the unstaged changes for the web console - a new diff will be regenerated once
+# we revendor the Web Console for OSE.
+git checkout -- pkg/assets/bindata.go pkg/assets/java/bindata.go
+
+# We need to retag, because the previous squash removed the latest tito tag.
+git tag "${last_tag}" HEAD
+
+# Rebase OSE on Origin
+git rebase upstream/master
+```
+
+New releases should be cut as they already are today (a mixture of git and tito). CVEs and bugfixes
+should continue to land in release branches as they already are today too. CVEs and important bugfixes
+will continue to be backported back to Origin manually but we may eventually build a script that will
+create git commits locally automatically, eg. hack/cherry-pick-from-ose.sh openshift/ose/pull/578.
+
+Common conflicts can rise when new changes step on branding or tooling code. Manual resolution is the
+only thing we can do in such cases but we can reduce the amount of conflicts by using build flags that
+will replace branding instances depending on the repository that is being built. We should also investigate
+whether we can stop checking in various code like tito diffs, generated docs, and the ose-images script.
+Avoiding those diffs will reduce even further the number of conflicts.
+
+In case of manual conflict resolution, or when a new [CARRY] commit needs to be introduced in master, or
+when an existing [CARRY] commit needs to be updated or dropped, changes will need to be force-pushed and
+the latest commits should always be as described above. The most common flow for ensuring commit order
+should be:
+```sh
+git remote set upstream git@github.com:openshift/origin
+git fetch upstream
+
+git add -A
+git commit -m "[CARRY] for a reason"
+
+# Rebase against the last Origin HEAD that was brought in during the previous rebase. This should open
+# the git editor of your choice and include the commits we are carrying on top of Origin. Reorder
+# appropriately to ensure the commit order described above is respected.
+git rebase -i $(git merge-base master upstream/master)
+git push origin master -f
```

Comment:
This requires that contributors have force-push rights to `master`. Can we use a flow by which the rebase script identifies if there are three consecutive commits as described above, and if there is anything else after it, rebases to squash them down or make them come before the last three that we require?